src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,284 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
+                            req_opts->cb, req_opts->rock, princ, &valid_san);
+    if (ret)
+        goto err;
+
+    if (!valid_san) {
+        pkiDebug("%s: did not find an acceptable SAN in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+        goto err;
+    }
+
+    /* Verify the client EKU. */
+    ret = verify_client_eku(context, req_opts->plgctx, req_opts->reqctx,
+                            &valid_eku);
+    if (ret)
+        goto err;
+
+    if (!valid_eku) {
+        pkiDebug("%s: did not find an acceptable EKU in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_INCONSISTENT_KEY_PURPOSE;
+    }
+
+err:
+    return ret;
+}
+
+static krb5_error_code
+certauth_pkinitsrv_initvt(krb5_context context, int maj_ver, int min_ver,
+                          krb5_plugin_vtable vtable)
+{
+    krb5_certauth_vtable vt;
+
+    if (maj_ver != 1)
+        return KRB5_PLUGIN_VER_NOTSUPP;
+    vt = (krb5_certauth_vtable)vtable;
+    vt->name = "pkinitsrv";
+    vt->authorize = pkinitsrv_authorize;
+    vt->init = NULL;
+    vt->fini = NULL;
+    vt->free_ind = NULL;
+    return 0;
+}
+
+/*
+ * Do certificate auth based on match expressions in the pkinit_cert_match
+ * attribute string.  An expression should be in the same form as those used for
+ * the pkinit_cert_match configuration option.  Multiple expressions are
+ * separated by ':'.  The certificate is authorized if all expressions are
+ * satisfied.
+ */
+static krb5_error_code
+dbmatch_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                  krb5_data cert, krb5_principal princ, void *opts,
+                  void *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    struct certauth_req_opts *req_opts = opts;
+    char *match = NULL, *matchcpy = NULL, *mp, *save;
+    krb5_boolean matched;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    ret = req_opts->cb->get_string(context, req_opts->rock,
+                                   "pkinit_cert_match", &match);
+    if (ret)
+        goto out;
+
+    /* Pass if there's no match expression. */
+    if (match == NULL) {
+        ret = KRB5_PLUGIN_NO_HANDLE;
+        goto out;
+    }
+
+    matchcpy = strdup(match);
+    req_opts->cb->free_string(context, req_opts->rock, match);
+    if (matchcpy == NULL) {
+        ret = ENOMEM;
+        goto out;
+    }
+
+    /* Decode certificate. */
+    ret = crypto_decode_der_cert(context, cert, req_opts->reqctx->cryptoctx);
+    if (ret)
+        goto out;
+
+    /* Check the certificate against the match expression. */
+    ret = pkinit_client_cert_match(context, req_opts->plgctx->cryptoctx,
+                                   req_opts->reqctx->cryptoctx, matchcpy,
+                                   &matched);
+    if (ret)
+        goto out;
+
+    if (!matched)
+        ret = KRB5KDC_ERR_CERTIFICATE_MISMATCH;
```

Comment:
I'm not sure if this should be a "no", but leave it alone for now; there's a reasonable argument that if the string attribute is present, it should be necessary as well as sufficient.